### PR TITLE
ARROW-10962: [FlightRPC][Java] fill in empty body buffer if needed

### PR DESCRIPTION
**Pull request by lidavidm**
Fixes compatibility with standard Protobuf implementations, which may omit serialization of an empty field. In the case of an empty record batch, this would cause the Java client to fail a precondition check since there would be no body buffer instead of a single empty body buffer.

### CODE REVIEW

**Comment by alamb** [comment]
Requesting review from @praveenbingo and @liyafan82 as they seem to have merged the last few Java PRs into the arrow repo. 

The background here is that we are working on flight integration tests in Rust (e.g. https://github.com/apache/arrow/pull/9049) and we found an edge case for flight integration.

**Comment by liyafan82** [comment]
Sorry for my late response. Will take a look late today.

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java [hunk 1]
```diff
@@ -317,6 +325,39 @@ private void test(BiConsumer<FlightClient, BufferAllocator> consumer) throws Exc
     }
   }
 
+  /** ARROW-10939: accept FlightData messages generated by Protobuf (which can omit empty fields). */
```
Comment (liyafan82) [inline java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java:328]: 10939 -> 10962?

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java [hunk 2]
```diff
@@ -317,6 +325,39 @@ private void test(BiConsumer<FlightClient, BufferAllocator> consumer) throws Exc
...
+  @Test
+  public void testProtobufCompat() throws Exception {
```
Comment (liyafan82) [inline java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java:330]: Compat -> Compact?

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java [hunk 3]
```diff
@@ -317,6 +325,39 @@ private void test(BiConsumer<FlightClient, BufferAllocator> consumer) throws Exc
...
+        }
+      }
+      final byte[] serializedMessage = baos.toByteArray();
+      final Flight.FlightData protobufData = Flight.FlightData.parseFrom(serializedMessage);
+      Assert.assertEquals(0, protobufData.getDataBody().size());
+      // Should not throw
+      final ArrowRecordBatch rb =
+          marshaller.parse(new ByteArrayInputStream(protobufData.toByteArray())).asRecordBatch();
+      Assert.assertEquals(rb.computeBodyLength(), 0);
+    }
```
Comment (liyafan82) [inline java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java:358]: nit: do you think we also need a test case for which a schema has an empty body?

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java (hunk 2 above)
Comment (lidavidm) [reply java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java]: Expanded to 'compatibility'

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java (hunk 1 above)
Comment (lidavidm) [reply java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java]: Good catch, fixed.

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java (hunk 3 above)
Comment (lidavidm) [reply java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java:358]: Yup, I added a test for this as well.

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java [hunk 4]
```diff
@@ -317,6 +325,71 @@ private void test(BiConsumer<FlightClient, BufferAllocator> consumer) throws Exc
...
+        }
+      }
+      final byte[] serializedMessage = baos.toByteArray();
+      final Flight.FlightData protobufData = Flight.FlightData.parseFrom(serializedMessage)
+          .toBuilder()
+          .setDataBody(ByteString.EMPTY)
+          .build();
+      Assert.assertEquals(0, protobufData.getDataBody().size());
+      // Should not throw
+      marshaller.parse(new ByteArrayInputStream(protobufData.toByteArray())).asSchema();
```
Comment (liyafan82) [inline java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java:389]: Maybe we need to assign the schema to an object and verify that its body is null?

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java [hunk 5]
```diff
@@ -317,6 +325,71 @@ private void test(BiConsumer<FlightClient, BufferAllocator> consumer) throws Exc
...
+  @Test
+  public void testProtobufSchemaCompatibility() throws Exception {
+    final Schema schema = new Schema(Collections.singletonList(Field.nullable("foo", new ArrowType.Int(32, true))));
+    try (final BufferAllocator allocator = new RootAllocator(Integer.MAX_VALUE)) {
+      final MethodDescriptor.Marshaller<ArrowMessage> marshaller = ArrowMessage.createMarshaller(allocator);
+      final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+      Flight.FlightDescriptor descriptor = FlightDescriptor.command(new byte[0]).toProtocol();
+      try (final ArrowMessage message = new ArrowMessage(descriptor, schema, new IpcOption())) {
+        Assert.assertEquals(ArrowMessage.HeaderType.SCHEMA, message.getMessageType());
+        try (final InputStream serialized = marshaller.stream(message)) {
```
Comment (liyafan82) [inline java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java:371]: We can extract a common method for the logic of writing the message to a ByteArrayOutputStream?

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java [hunk 6]
```diff
@@ -317,6 +325,71 @@ private void test(BiConsumer<FlightClient, BufferAllocator> consumer) throws Exc
...
+
+  /** ARROW-10962: accept FlightData messages generated by Protobuf (which can omit empty fields). */
+  @Test
+  public void testProtobufSchemaCompatibility() throws Exception {
+    final Schema schema = new Schema(Collections.singletonList(Field.nullable("foo", new ArrowType.Int(32, true))));
+    try (final BufferAllocator allocator = new RootAllocator(Integer.MAX_VALUE)) {
+      final MethodDescriptor.Marshaller<ArrowMessage> marshaller = ArrowMessage.createMarshaller(allocator);
+      final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+      Flight.FlightDescriptor descriptor = FlightDescriptor.command(new byte[0]).toProtocol();
+      try (final ArrowMessage message = new ArrowMessage(descriptor, schema, new IpcOption())) {
```
Comment (liyafan82) [inline java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java:369]: Do we need to check the message body here?

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java (hunk 4 above)
Comment (lidavidm) [reply java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java]: The schema itself has no body. I've added a test to check the parsed message.

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java (hunk 6 above)
Comment (lidavidm) [reply java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java:391]: I've added assertions here.

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java (hunk 5 above)
Comment (lidavidm) [reply java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java]: Done.

**Review by liyafan82 — APPROVED** [review]
LGTM. Will merge soon if there are no more comments.

**Comment by alamb** [comment]
https://github.com/apache/arrow/pull/9049 is still blocked on getting this PR merged. Any chance you can do so @liyafan82 ?

**Comment by liyafan82** [comment]
Merging. Thanks for your effort. @lidavidm @alamb

**Comment by alamb** [comment]
🎉  thanks @liyafan82  -- I didn't do anything, it was all @lidavidm . Thanks @lidavidm for the efforts!

**Pull request closed without merging**